File: src/main/golang/leetcode/editor/cn/[15]3Sum.go
package lc

import "sort"

//leetcode submit region begin(Prohibit modification and deletion)
func threeSum(nums []int) (ans [][]int) {
	sort.Ints(nums)
	n := len(nums)
	findTarget := func(start, target int) (ans [][]int) {
		l, r := start, n-1
		for l < r {
			sum := nums[l] + nums[r]
			left, right := nums[l], nums[r]
			if sum < target {
				for l < r && nums[l] == left {
					l++
				}
			} else if sum > target {
				for l < r && nums[r] == right {
					r--
				}
			} else {
				ans = append(ans, []int{nums[l], nums[r]})
				for l < r && nums[l] == left {
					l++
				}
				for l < r && nums[r] == right {
					r--
				}
			}
		}
		return
	}

	for i := 0; i < n; i++ {
		res := findTarget(i+1, -nums[i])
		for _, slice := range res {
			ans = append(ans, []int{nums[i], slice[0], slice[1]})
		}
		for i < n-2 && nums[i+1] == nums[i] {
			i++
		}
	}
	return
}

//leetcode submit region end(Prohibit modification and deletion)

File: src/main/golang/leetcode/editor/cn/[15]3Sum.py
# leetcode submit region begin(Prohibit modification and deletion)
from typing import List


class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        n = len(nums)
        ans = []
        previous = -10 ** 5 - 1
        for i in range(0, n - 2):
            target = nums[i]
            if target == previous:
                continue
            previous = target
            l, r = i + 1, n - 1
            while l < r:
                left, right = nums[l], nums[r]
                if left + right == -target:
                    ans.append([target, left, right])
                    while l < r and nums[l] == left: l += 1
                    while l < r and nums[r] == right: r -= 1
                elif left + right > -target:
                    while l < r and nums[r] == right: r -= 1
                else:
                    while l < r and nums[l] == left: l += 1
        print(ans)
        return ans
# leetcode submit region end(Prohibit modification and deletion)
